File: src/systems/FarmSystem.ts
// Farming: sow a seed into a Farm Plot bed, come back when it is ripe.
//
// Crops grow on **wall-clock time**, stored as a plant timestamp. That is the
// whole design: there is no tick loop to run, no catch-up pass on load, and no
// way for growth to drift out of step with the save — a bed sown before you close
// the tab is simply ripe when `Date.now()` says it is. Every other production
// system in the game needed an explicit offline calculation; this one gets it for
// free, which is also why it cannot be double-paid the way offline gathering was.
//
// Bed count comes from Farm Plot *levels*, matching every other passive effect
// (see BuildSystem.levels), so upgrading a plot adds beds.
import type { GameState } from "../state/GameState";
import { addItem, countItem, removeItem, isBulk, storedAmount } from "../components/Inventory";
import { addMasteryXp, masteryLevel } from "../components/Skills";
import { levelFromXp } from "../data/XPTable";
import { SEEDS, SEED_IDS, growthAt, growthLabel, isRipe, type FarmPlot, type SeedDef } from "../data/Farming";
import { ITEMS } from "../data/Items";

export type PlantFailure =
  | "no_bed"        // every bed is occupied
  | "no_seed"       // none of that seed carried
  | "level"         // farming level too low
  | "unknown_seed";

export type HarvestFailure = "empty" | "unripe" | "inventory_full";

/** One bed as the UI sees it. */
export interface FarmBedRow {
  index: number;
  seedId: string | null;
  name: string;
  icon: string;
  /** 0..1; 1 = ripe. */
  growth: number;
  label: string;
  ripe: boolean;
}

export interface FarmSnapshot {
  beds: FarmBedRow[];
  /** Seeds the player carries, with whether they can be sown right now. */
  seeds: { id: string; name: string; qty: number; levelReq: number; unlocked: boolean }[];
  level: number;
  /** 0 when no Farm Plot is built — the panel explains rather than showing nothing. */
  bedCount: number;
}

export class FarmSystem {
  private state: GameState;
  /** Beds available = total Farm Plot levels. Injected so BuildSystem stays the
   *  single authority on what the settlement has. */
  private bedProvider: () => number;

  constructor(state: GameState, bedProvider: () => number) {
    this.state = state;
    this.bedProvider = bedProvider;
  }

  get bedCount(): number { return Math.max(0, this.bedProvider()); }

  private get plots(): (FarmPlot | null)[] { return this.state.town.farm.plots; }

  /** Farming level from XP. */
  get level(): number { return levelFromXp(this.state.player.skills.farming.xp); }

  /**
   * Trim or extend the bed array to match the buildings.
   *
   * Beds are addressed by index, so the array has to track the building count —
   * but shrinking it must not silently bin a growing crop, so only empty trailing
   * beds are removed.
   */
  private sync(): void {
    const want = this.bedCount;
    const p = this.plots;
    while (p.length < want) p.push(null);
    while (p.length > want && p[p.length - 1] === null) p.pop();
  }

  /** Index of the first free bed, or -1. */
  private freeBed(): number {
    this.sync();
    return this.plots.findIndex((b) => b === null);
  }

  plant(seedId: string): { ok: true; bed: number } | { ok: false; reason: PlantFailure } {
    const def: SeedDef | undefined = SEEDS[seedId];
    if (!def) return { ok: false, reason: "unknown_seed" };
    if (this.level < def.levelReq) return { ok: false, reason: "level" };
    if (countItem(this.state.player.inventory, seedId) < 1) return { ok: false, reason: "no_seed" };
    const bed = this.freeBed();
    if (bed < 0) return { ok: false, reason: "no_bed" };
    removeItem(this.state.player.inventory, seedId, 1);
    this.plots[bed] = { seedId, plantedAt: Date.now() };
    return { ok: true, bed };
  }

  harvest(bed: number): { ok: true; itemId: string; amount: number; xp: number } | { ok: false; reason: HarvestFailure } {
    this.sync();
    const plot = this.plots[bed];
    if (!plot) return { ok: false, reason: "empty" };
    const now = Date.now();
    if (!isRipe(plot, now)) return { ok: false, reason: "unripe" };
    const def = SEEDS[plot.seedId];
    if (!def) { this.plots[bed] = null; return { ok: false, reason: "empty" }; }

    const inv = this.state.player.inventory;
    // Mastery raises the yield FLOOR rather than adding a separate bonus roll: at
    // mastery 1 the harvest spans the crop's full range, at 99 it is always the
    // maximum. One knob, and the range in the wiki stays literally true.
    const m = masteryLevel(this.state.player.skills.farming.mastery[def.masteryKey] || 0);
    const span = def.produce.max - def.produce.min;
    const lo = def.produce.min + Math.floor(span * ((m - 1) / 98));
    const yielded = lo + Math.floor(Math.random() * (def.produce.max - lo + 1));

    if (isBulk(def.produce.itemId) && storedAmount(inv) + yielded > inv.storageCap) {
      return { ok: false, reason: "inventory_full" };
    }

    this.plots[bed] = null;
    addItem(inv, def.produce.itemId, yielded);
    this.state.player.skills.farming.xp += def.xp;
    addMasteryXp(this.state.player.skills, "farming", def.masteryKey, yielded);
    this.state.collectionLog.add(def.produce.itemId);
    return { ok: true, itemId: def.produce.itemId, amount: yielded, xp: def.xp };
  }

  /** Harvest every ripe bed. Returns what was gathered, for one summary toast. */
  harvestAll(): { itemId: string; amount: number }[] {
    this.sync();
    const got: { itemId: string; amount: number }[] = [];
    for (let i = 0; i < this.plots.length; i++) {
      const r = this.harvest(i);
      if (r.ok) {
        const prev = got.find((g) => g.itemId === r.itemId);
        if (prev) prev.amount += r.amount;
        else got.push({ itemId: r.itemId, amount: r.amount });
      }
    }
    return got;
  }

  /** How many beds are ready right now — drives the Village button's badge. */
  ripeCount(): number {
    const now = Date.now();
    this.sync();
    return this.plots.filter((p): p is FarmPlot => !!p && isRipe(p, now)).length;
  }

  snapshot(): FarmSnapshot {
    this.sync();
    const now = Date.now();
    const level = this.level;
    return {
      bedCount: this.bedCount,
      level,
      beds: this.plots.map((p, index) => {
        if (!p) return { index, seedId: null, name: "Empty bed", icon: "🟫", growth: 0, label: "Ready to sow", ripe: false };
        const def = SEEDS[p.seedId];
        return {
          index,
          seedId: p.seedId,
          name: def ? (ITEMS[def.produce.itemId]?.name ?? def.produce.itemId) : p.seedId,
          icon: def && isRipe(p, now) ? "🌾" : "🌱",
          growth: growthAt(p, now),
          label: growthLabel(p, now),
          ripe: isRipe(p, now),
        };
      }),
      seeds: SEED_IDS.map((id) => ({
        id,
        name: SEEDS[id].name,
        qty: countItem(this.state.player.inventory, id),
        levelReq: SEEDS[id].levelReq,
        unlocked: level >= SEEDS[id].levelReq,
      })).filter((s) => s.qty > 0 || s.unlocked),
    };
  }
}

File: unity/Assets/Isoperia/Core/Runtime/Systems/FarmSystem.cs
using System;
using System.Collections.Generic;
using Isoperia.Core.Components;
using Isoperia.Core.Content;
using Isoperia.Core.Data;
using Isoperia.Core.Save;
using Isoperia.Core.Sim;
using Isoperia.Core.State;

namespace Isoperia.Core.Systems
{
    public enum PlantFailure { None, UnknownSeed, Level, NoSeed, NoBed }

    public enum HarvestFailure { None, Empty, Unripe, InventoryFull }

    public sealed class HarvestResult
    {
        public bool Ok;
        public HarvestFailure Reason;
        public string ItemId;
        public int Amount;
        public int Xp;
    }

    /// <summary>
    /// Farming. Port of <c>src/systems/FarmSystem.ts</c>.
    ///
    /// CROPS GROW ON WALL-CLOCK TIME, stored as a plant timestamp. That is the
    /// whole design, and it is why this system needs no offline pass: a bed sown
    /// before the tab closed is simply ripe when the clock says so. Every other
    /// production system needed an explicit offline calculation and could be
    /// double-paid; this one cannot. Do not "improve" it into a tick accumulator.
    ///
    /// The clock is injected rather than read from the system, so growth can be
    /// tested without waiting five minutes.
    /// </summary>
    public sealed class FarmSystem
    {
        private readonly GameState _state;
        private readonly ContentDatabase _content;
        private readonly IRandom _rng;
        private readonly Func<long> _now;
        private readonly Func<int> _beds;

        public FarmSystem(GameState state, ContentDatabase content, IRandom rng,
                          Func<long> nowMsProvider, Func<int> bedProvider)
        {
            _state = state ?? throw new ArgumentNullException(nameof(state));
            _content = content ?? throw new ArgumentNullException(nameof(content));
            _rng = rng ?? throw new ArgumentNullException(nameof(rng));
            _now = nowMsProvider ?? throw new ArgumentNullException(nameof(nowMsProvider));
            _beds = bedProvider ?? throw new ArgumentNullException(nameof(bedProvider));
        }

        /// <summary>
        /// Beds available = total FARM_PLOT *levels*, not the number of plots.
        /// Every passive effect scales on levels, so a level-3 plot gives three
        /// beds; the alternative made upgrading cost 2x then 3x the materials and
        /// buy nothing.
        /// </summary>
        public static int BedsFrom(GameState state)
        {
            int n = 0;
            foreach (TownBuilding b in state.Town.Buildings)
                if (b.Type == "FARM_PLOT") n += Math.Max(1, b.Level);
            return n;
        }

        public int BedCount => Math.Max(0, _beds());

        public int Level => _state.Player.Skills.LevelOf(Skills.Farming);

        private List<FarmPlot> Plots => _state.Town.FarmPlots;

        /// <summary>
        /// Match the bed array to the buildings.
        ///
        /// Only EMPTY TRAILING beds are removed. Shrinking past a growing crop
        /// would silently bin it, and a demolished plot must not destroy what is
        /// already in the ground.
        /// </summary>
        private void Sync()
        {
            int want = BedCount;
            List<FarmPlot> p = Plots;

            while (p.Count < want) p.Add(null);
            while (p.Count > want && p.Count > 0 && p[p.Count - 1] == null) p.RemoveAt(p.Count - 1);
        }

        private JsonValue Seed(string id)
        {
            JsonValue s = _content.Seeds[id];
            return s.IsNull ? null : s;
        }

        /// <summary>0..1, clamped. 1 means ripe.</summary>
        public double GrowthAt(FarmPlot plot, long now)
        {
            if (plot == null) return 0;

            JsonValue def = Seed(plot.SeedId);
            double grow = def == null ? 0 : def["growMs"].AsNumber(0);
            if (def == null || grow <= 0) return 1;

            return Math.Max(0, Math.Min(1, (now - plot.PlantedAt) / grow));
        }

        public bool IsRipe(FarmPlot plot, long now) => GrowthAt(plot, now) >= 1;

        public string GrowthLabel(FarmPlot plot, long now)
        {
            JsonValue def = plot == null ? null : Seed(plot.SeedId);
            if (def == null) return "—";

            double leftMs = def["growMs"].AsNumber(0) - (now - plot.PlantedAt);
            if (leftMs <= 0) return "Ripe";

            long mins = (long)Math.Ceiling(leftMs / 60000.0);
            return mins >= 60 ? $"{mins / 60}h {mins % 60}m left" : $"{mins}m left";
        }

        /// <summary>Index of the first free bed, or -1.</summary>
        private int FreeBed()
        {
            Sync();
            for (int i = 0; i < Plots.Count; i++) if (Plots[i] == null) return i;
            return -1;
        }

        public bool TryPlant(string seedId, out int bed, out PlantFailure reason)
        {
            bed = -1;

            JsonValue def = Seed(seedId);
            if (def == null) { reason = PlantFailure.UnknownSeed; return false; }

            if (Level < (int)def["levelReq"].AsNumber(1)) { reason = PlantFailure.Level; return false; }
            if (_state.Player.Inventory.Count(seedId) < 1) { reason = PlantFailure.NoSeed; return false; }

            int free = FreeBed();
            if (free < 0) { reason = PlantFailure.NoBed; return false; }

            _state.Player.Inventory.Remove(seedId, 1);
            Plots[free] = new FarmPlot { SeedId = seedId, PlantedAt = _now() };

            bed = free;
            reason = PlantFailure.None;
            return true;
        }

        /// <summary>
        /// Harvest one bed.
        ///
        /// Takes exactly ONE random draw, and only once every gate has passed —
        /// so a failed harvest does not advance the stream. Callers that harvest
        /// several beds therefore consume one draw per SUCCESSFUL bed.
        /// </summary>
        public HarvestResult Harvest(int bed)
        {
            Sync();

            if (bed < 0 || bed >= Plots.Count || Plots[bed] == null)
                return new HarvestResult { Reason = HarvestFailure.Empty };

            FarmPlot plot = Plots[bed];
            long now = _now();

            if (!IsRipe(plot, now))
                return new HarvestResult { Reason = HarvestFailure.Unripe };

            JsonValue def = Seed(plot.SeedId);
            if (def == null)
            {
                // The seed left the content tables while a crop was in the
                // ground. Clear the bed rather than stranding it forever.
                Plots[bed] = null;
                return new HarvestResult { Reason = HarvestFailure.Empty };
            }

            string masteryKey = def["masteryKey"].AsString("");
            JsonValue produce = def["produce"];
            string itemId = produce["itemId"].AsString(null);
            int min = (int)produce["min"].AsNumber(1);
            int max = (int)produce["max"].AsNumber(1);

            // Mastery raises the yield FLOOR rather than adding a bonus roll: at
            // mastery 1 the harvest spans the crop's full range, at 99 it is
            // always the maximum. One knob, and the range printed in the wiki
            // stays literally true.
            _state.Player.Skills.Get(Skills.Farming).Mastery.TryGetValue(masteryKey, out double mXp);
            int m = MasteryTable.LevelFromXp(mXp);

            int span = max - min;
            int lo = min + (int)Math.Floor(span * ((m - 1) / 98.0));
            int yielded = lo + (int)Math.Floor(_rng.Next() * (max - lo + 1));

            InventoryComponent inv = _state.Player.Inventory;

            // Checked BEFORE the bed is cleared, so a full bag leaves the crop in
            // the ground rather than destroying it. (LabourSystem.Claim does the
            // opposite and loses the overflow — that asymmetry is in the
            // TypeScript too.)
            if (inv.IsBulkItem(itemId) && inv.StoredAmount() + yielded > inv.StorageCap)
                return new HarvestResult { Reason = HarvestFailure.InventoryFull };

            Plots[bed] = null;
            inv.Add(itemId, yielded);

            int xp = (int)def["xp"].AsNumber(0);
            _state.Player.Skills.AddXp(Skills.Farming, xp);
            _state.Player.Skills.AddMasteryXp(Skills.Farming, masteryKey, yielded);
            _state.CollectionLog.Add(itemId);

            return new HarvestResult
            {
                Ok = true, Reason = HarvestFailure.None,
                ItemId = itemId, Amount = yielded, Xp = xp,
            };
        }

        /// <summary>Harvest every ripe bed, totalled per item for one summary.</summary>
        public List<KeyValuePair<string, int>> HarvestAll()
        {
            Sync();

            var got = new List<KeyValuePair<string, int>>();

            for (int i = 0; i < Plots.Count; i++)
            {
                HarvestResult r = Harvest(i);
                if (!r.Ok) continue;

                int at = got.FindIndex(g => g.Key == r.ItemId);
                if (at >= 0) got[at] = new KeyValuePair<string, int>(r.ItemId, got[at].Value + r.Amount);
                else got.Add(new KeyValuePair<string, int>(r.ItemId, r.Amount));
            }

            return got;
        }

        /// <summary>Beds ready right now — drives the Village button's badge.</summary>
        public int RipeCount()
        {
            Sync();
            long now = _now();

            int n = 0;
            foreach (FarmPlot p in Plots) if (p != null && IsRipe(p, now)) n++;
            return n;
        }
    }
}
